dali/pipeline/operators/detection/box_encoder.cc

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+
+#include "dali/pipeline/operators/detection/box_encoder.h"
+
+namespace dali {
+namespace detail {
+using Anchor = BoxEncoder<CPUBackend>::Anchor;
+using BBox = BoxEncoder<CPUBackend>::BBox;
+
+float Iou(const Anchor a1, const Anchor a2) {
+  float l = std::max(a1.x, a2.x);
+  float t = std::max(a1.y, a2.y);
+  float r = std::min(a1.z, a2.z);
+  float b = std::min(a1.w, a2.w);
+
+  float width = r - l;
+  float height = b - t;
+
+  if (width <= 0.f || height <= 0.f) return 0.f;
+
+  float intersection = width * height;
+  float area1 = (a1.w - a1.y) * (a1.z - a1.x);
+  float area2 = (a2.w - a2.y) * (a2.z - a2.x);
+
+  return intersection / (area1 + area2 - intersection);
+}
+
+vector<float> CalculateIous(const Anchor *anchors, const BBox *bboxes, int N, int M) {
+  vector<float> ious(N * M);
```

Comment:
If it is fixed size how about using just std::array?

Reply:
It is not fixed sized, as N is the number of bounding boxes in current sample. I was trying to follow naming convention from mlperf implementation, thus capital N and M, but since it is misleading changed it to more descriptive names (```num_boxes``` and ```NumAnchors()```)

Reply:
Ok, I missed that.